File: sdks/ios/Sources/TaskirX/Services/Services.swift
import Foundation

// MARK: - Auth Service

public class AuthService {
    private let requestManager: RequestManager
    private var currentUser: User?
    private let userLock = NSLock()
    
    init(requestManager: RequestManager) {
        self.requestManager = requestManager
    }
    
    public func register(email: String, password: String, name: String, company: String? = nil) async throws -> AuthResponse {
        let request = RegisterRequest(email: email, password: password, name: name, company: company)
        let response: AuthResponse = try await requestManager.post("/auth/register", body: request)
        
        userLock.lock()
        defer { userLock.unlock() }
        currentUser = response.user
        requestManager.setAuthToken(response.token)
        
        return response
    }
    
    public func login(email: String, password: String) async throws -> AuthResponse {
        let request = LoginRequest(email: email, password: password)
        let response: AuthResponse = try await requestManager.post("/auth/login", body: request)
        
        userLock.lock()
        defer { userLock.unlock() }
        currentUser = response.user
        requestManager.setAuthToken(response.token)
        
        return response
    }
    
    public func logout() async throws -> [String: AnyCodable] {
        requestManager.clearAuthToken()
        userLock.lock()
        defer { userLock.unlock() }
        currentUser = nil
        return [:]
    }
    
    public func getProfile() async throws -> User {
        let user: User = try await requestManager.get("/auth/profile")
        userLock.lock()
        defer { userLock.unlock() }
        currentUser = user
        return user
    }
    
    public func refreshToken(refreshToken: String) async throws -> AuthResponse {
        let request = ["refreshToken": AnyCodable.string(refreshToken)]
        let response: AuthResponse = try await requestManager.post("/auth/refresh", body: request)
        requestManager.setAuthToken(response.token)
        return response
    }
}

// MARK: - Campaign Service

public class CampaignService {
    private let requestManager: RequestManager
    
    init(requestManager: RequestManager) {
        self.requestManager = requestManager
    }
    
    public func create(name: String, budget: Double, startDate: String, 
                      endDate: String, targetAudience: [String: AnyCodable]) async throws -> Campaign {
        let request = CampaignCreateRequest(name: name, budget: budget, startDate: startDate, 
                                           endDate: endDate, targetAudience: targetAudience)
        return try await requestManager.post("/campaigns", body: request)
    }
    
    public func list(limit: Int = 50, offset: Int = 0) async throws -> [Campaign] {
        let endpoint = "/campaigns?limit=\(limit)&offset=\(offset)"
        return try await requestManager.get(endpoint)
    }
    
    public func get(id: String) async throws -> Campaign {
        return try await requestManager.get("/campaigns/\(id)")
    }
    
    public func update(id: String, name: String? = nil, budget: Double? = nil) async throws -> Campaign {
        var updates: [String: AnyCodable] = [:]
        if let name = name { updates["name"] = .string(name) }
        if let budget = budget { updates["budget"] = .double(budget) }
        return try await requestManager.put("/campaigns/\(id)", body: updates)
    }
    
    public func delete(id: String) async throws -> [String: AnyCodable] {
        return try await requestManager.delete("/campaigns/\(id)")
    }
    
    public func pause(id: String) async throws -> Campaign {
        return try await requestManager.put("/campaigns/\(id)/pause")
    }
    
    public func resume(id: String) async throws -> Campaign {
        return try await requestManager.put("/campaigns/\(id)/resume")
    }
}

// MARK: - Analytics Service

public class AnalyticsService {
    private let requestManager: RequestManager
    
    init(requestManager: RequestManager) {
        self.requestManager = requestManager
    }
    
    public func realtime() async throws -> Analytics {
        return try await requestManager.get("/analytics/realtime")
    }
    
    public func campaign(id: String) async throws -> Analytics {
        return try await requestManager.get("/analytics/campaigns/\(id)")
    }
    
    public func breakdown(type: String) async throws -> [[String: AnyCodable]] {
        return try await requestManager.get("/analytics/breakdown?type=\(type)")
    }
    
    public func dashboard() async throws -> [String: AnyCodable] {
        return try await requestManager.get("/analytics/dashboard")
    }
}

// MARK: - Bidding Service

public class BiddingService {
    private let requestManager: RequestManager
    
    init(requestManager: RequestManager) {
        self.requestManager = requestManager
    }
    
    public func submitBid(campaignId: String, adSlotId: String, amount: Double, currency: String = "USD") async throws -> Bid {
        let request = BidSubmitRequest(campaignId: campaignId, adSlotId: adSlotId, amount: amount, currency: currency)
        return try await requestManager.post("/bids", body: request)
    }
    
    public func recommendations() async throws -> [[String: AnyCodable]] {
        return try await requestManager.get("/bids/recommendations")
    }
    
    public func list(limit: Int = 50) async throws -> [Bid] {
        return try await requestManager.get("/bids?limit=\(limit)")
    }
    
    public func get(id: String) async throws -> Bid {
        return try await requestManager.get("/bids/\(id)")
    }
    
    public func stats() async throws -> [String: AnyCodable] {
        return try await requestManager.get("/bids/stats")
    }
}

// MARK: - Ad Service

public class AdService {
    private let requestManager: RequestManager
    
    init(requestManager: RequestManager) {
        self.requestManager = requestManager
    }
    
    public func create(campaignId: String, placement: String, imageUrl: String, 
                      clickUrl: String, dimensions: String) async throws -> Ad {
        let request = AdCreateRequest(campaignId: campaignId, placement: placement, 
                                     imageUrl: imageUrl, clickUrl: clickUrl, dimensions: dimensions)
        return try await requestManager.post("/ads", body: request)
    }
    
    public func list(campaignId: String, limit: Int = 50) async throws -> [Ad] {
        return try await requestManager.get("/ads?campaignId=\(campaignId)&limit=\(limit)")
    }
    
    public func get(id: String) async throws -> Ad {
        return try await requestManager.get("/ads/\(id)")
    }
    
    public func update(id: String, placement: String? = nil) async throws -> Ad {
        var updates: [String: AnyCodable] = [:]
        if let placement = placement { updates["placement"] = .string(placement) }
        return try await requestManager.put("/ads/\(id)", body: updates)
    }
    
    public func delete(id: String) async throws -> [String: AnyCodable] {
        return try await requestManager.delete("/ads/\(id)")
    }
}

// MARK: - Webhook Service

public class WebhookService {
    private let requestManager: RequestManager
    private var eventHandlers: [String: [(WebhookEvent) -> Void]] = [:]
    private let handlerLock = NSLock()
    
    init(requestManager: RequestManager) {
        self.requestManager = requestManager
    }
    
    public func subscribe(url: String, events: [String]) async throws -> Webhook {
        let request = WebhookCreateRequest(url: url, events: events)
        return try await requestManager.post("/webhooks", body: request)
    }
    
    public func list(limit: Int = 50) async throws -> [Webhook] {
        return try await requestManager.get("/webhooks?limit=\(limit)")
    }
    
    public func get(id: String) async throws -> Webhook {
        return try await requestManager.get("/webhooks/\(id)")
    }
    
    public func update(id: String, active: Bool? = nil) async throws -> Webhook {
        var updates: [String: AnyCodable] = [:]
        if let active = active { updates["active"] = .bool(active) }
        return try await requestManager.put("/webhooks/\(id)", body: updates)
    }
    
    public func delete(id: String) async throws -> [String: AnyCodable] {
        return try await requestManager.delete("/webhooks/\(id)")
    }
    
    public func test(id: String) async throws -> [String: AnyCodable] {
        return try await requestManager.post("/webhooks/\(id)/test")
    }
    
    public func getLogs(id: String, limit: Int = 50) async throws -> [[String: AnyCodable]] {
        return try await requestManager.get("/webhooks/\(id)/logs?limit=\(limit)")
    }
    
    public func onEvent(_ type: String, handler: @escaping (WebhookEvent) -> Void) {
        handlerLock.lock()
        defer { handlerLock.unlock() }
        if eventHandlers[type] == nil {
            eventHandlers[type] = []
        }
        eventHandlers[type]?.append(handler)
    }
    
    public func offEvent(_ type: String) {
        handlerLock.lock()
        defer { handlerLock.unlock() }
        eventHandlers[type] = nil
    }
    
    public func handleEvent(_ event: WebhookEvent) {
        handlerLock.lock()
        defer { handlerLock.unlock() }
        eventHandlers[event.type]?.forEach { $0(event) }
    }
}

File: sdks/react-native/src/services/Services.ts
// Services for TaskirX React Native SDK

import { RequestManager } from '../network/RequestManager';
import type {
  AuthResponse,
  LoginRequest,
  RegisterRequest,
  User,
  Campaign,
  CampaignCreateRequest,
  Bid,
  BidSubmitRequest,
  Analytics,
  Ad,
  AdCreateRequest,
  Webhook,
  WebhookCreateRequest,
  WebhookEvent,
} from '../types';

// Auth Service
export class AuthService {
  private requestManager: RequestManager;
  private currentUser: User | null = null;

  constructor(requestManager: RequestManager) {
    this.requestManager = requestManager;
  }

  async register(email: string, password: string, name: string, company?: string): Promise<AuthResponse> {
    const request: RegisterRequest = { email, password, name, company };
    const response = await this.requestManager.post<AuthResponse>('/auth/register', request);
    this.currentUser = response.user;
    this.requestManager.setAuthToken(response.token);
    return response;
  }

  async login(email: string, password: string): Promise<AuthResponse> {
    const request: LoginRequest = { email, password };
    const response = await this.requestManager.post<AuthResponse>('/auth/login', request);
    this.currentUser = response.user;
    this.requestManager.setAuthToken(response.token);
    return response;
  }

  async logout(): Promise<void> {
    this.requestManager.clearAuthToken();
    this.currentUser = null;
  }

  async getProfile(): Promise<User> {
    const user = await this.requestManager.get<User>('/auth/profile');
    this.currentUser = user;
    return user;
  }

  async refreshToken(refreshToken: string): Promise<AuthResponse> {
    const response = await this.requestManager.post<AuthResponse>('/auth/refresh', { refreshToken });
    this.requestManager.setAuthToken(response.token);
    return response;
  }
}

// Campaign Service
export class CampaignService {
  private requestManager: RequestManager;

  constructor(requestManager: RequestManager) {
    this.requestManager = requestManager;
  }

  async create(name: string, budget: number, startDate: string, endDate: string, targetAudience: Record<string, any>): Promise<Campaign> {
    const request: CampaignCreateRequest = { name, budget, startDate, endDate, targetAudience };
    return this.requestManager.post<Campaign>('/campaigns', request);
  }

  async list(limit: number = 50, offset: number = 0): Promise<Campaign[]> {
    return this.requestManager.get<Campaign[]>(`/campaigns?limit=${limit}&offset=${offset}`);
  }

  async get(id: string): Promise<Campaign> {
    return this.requestManager.get<Campaign>(`/campaigns/${id}`);
  }

  async update(id: string, name?: string, budget?: number): Promise<Campaign> {
    const updates: Record<string, any> = {};
    if (name) updates.name = name;
    if (budget) updates.budget = budget;
    return this.requestManager.put<Campaign>(`/campaigns/${id}`, updates);
  }

  async delete(id: string): Promise<Record<string, any>> {
    return this.requestManager.delete<Record<string, any>>(`/campaigns/${id}`);
  }

  async pause(id: string): Promise<Campaign> {
    return this.requestManager.put<Campaign>(`/campaigns/${id}/pause`);
  }

  async resume(id: string): Promise<Campaign> {
    return this.requestManager.put<Campaign>(`/campaigns/${id}/resume`);
  }
}

// Analytics Service
export class AnalyticsService {
  private requestManager: RequestManager;

  constructor(requestManager: RequestManager) {
    this.requestManager = requestManager;
  }

  async realtime(): Promise<Analytics> {
    return this.requestManager.get<Analytics>('/analytics/realtime');
  }

  async campaign(id: string): Promise<Analytics> {
    return this.requestManager.get<Analytics>(`/analytics/campaigns/${id}`);
  }

  async breakdown(type: string): Promise<Record<string, any>[]> {
    return this.requestManager.get<Record<string, any>[]>(`/analytics/breakdown?type=${type}`);
  }

  async dashboard(): Promise<Record<string, any>> {
    return this.requestManager.get<Record<string, any>>('/analytics/dashboard');
  }
}

// Bidding Service
export class BiddingService {
  private requestManager: RequestManager;

  constructor(requestManager: RequestManager) {
    this.requestManager = requestManager;
  }

  async submitBid(campaignId: string, adSlotId: string, amount: number, currency: string = 'USD'): Promise<Bid> {
    const request: BidSubmitRequest = { campaignId, adSlotId, amount, currency };
    return this.requestManager.post<Bid>('/bids', request);
  }

  async recommendations(): Promise<Record<string, any>[]> {
    return this.requestManager.get<Record<string, any>[]>('/bids/recommendations');
  }

  async list(limit: number = 50): Promise<Bid[]> {
    return this.requestManager.get<Bid[]>(`/bids?limit=${limit}`);
  }

  async get(id: string): Promise<Bid> {
    return this.requestManager.get<Bid>(`/bids/${id}`);
  }

  async stats(): Promise<Record<string, any>> {
    return this.requestManager.get<Record<string, any>>('/bids/stats');
  }
}

// Ad Service
export class AdService {
  private requestManager: RequestManager;

  constructor(requestManager: RequestManager) {
    this.requestManager = requestManager;
  }

  async create(campaignId: string, placement: string, imageUrl: string, clickUrl: string, dimensions: string): Promise<Ad> {
    const request: AdCreateRequest = { campaignId, placement, imageUrl, clickUrl, dimensions };
    return this.requestManager.post<Ad>('/ads', request);
  }

  async list(campaignId: string, limit: number = 50): Promise<Ad[]> {
    return this.requestManager.get<Ad[]>(`/ads?campaignId=${campaignId}&limit=${limit}`);
  }

  async get(id: string): Promise<Ad> {
    return this.requestManager.get<Ad>(`/ads/${id}`);
  }

  async update(id: string, placement?: string): Promise<Ad> {
    const updates: Record<string, any> = {};
    if (placement) updates.placement = placement;
    return this.requestManager.put<Ad>(`/ads/${id}`, updates);
  }

  async delete(id: string): Promise<Record<string, any>> {
    return this.requestManager.delete<Record<string, any>>(`/ads/${id}`);
  }
}

// Webhook Service
export class WebhookService {
  private requestManager: RequestManager;
  private eventHandlers: Map<string, Array<(event: WebhookEvent) => void>> = new Map();

  constructor(requestManager: RequestManager) {
    this.requestManager = requestManager;
  }

  async subscribe(url: string, events: string[]): Promise<Webhook> {
    const request: WebhookCreateRequest = { url, events };
    return this.requestManager.post<Webhook>('/webhooks', request);
  }

  async list(limit: number = 50): Promise<Webhook[]> {
    return this.requestManager.get<Webhook[]>(`/webhooks?limit=${limit}`);
  }

  async get(id: string): Promise<Webhook> {
    return this.requestManager.get<Webhook>(`/webhooks/${id}`);
  }

  async update(id: string, active?: boolean): Promise<Webhook> {
    const updates: Record<string, any> = {};
    if (active !== undefined) updates.active = active;
    return this.requestManager.put<Webhook>(`/webhooks/${id}`, updates);
  }

  async delete(id: string): Promise<Record<string, any>> {
    return this.requestManager.delete<Record<string, any>>(`/webhooks/${id}`);
  }

  async test(id: string): Promise<Record<string, any>> {
    return this.requestManager.post<Record<string, any>>(`/webhooks/${id}/test`);
  }

  async getLogs(id: string, limit: number = 50): Promise<Record<string, any>[]> {
    return this.requestManager.get<Record<string, any>[]>(`/webhooks/${id}/logs?limit=${limit}`);
  }

  onEvent(type: string, handler: (event: WebhookEvent) => void): void {
    if (!this.eventHandlers.has(type)) {
      this.eventHandlers.set(type, []);
    }
    this.eventHandlers.get(type)!.push(handler);
  }

  offEvent(type: string): void {
    this.eventHandlers.delete(type);
  }

  handleEvent(event: WebhookEvent): void {
    const handlers = this.eventHandlers.get(event.type);
    if (handlers) {
      handlers.forEach(handler => handler(event));
    }
  }
}
